src/main/java/org/myrobotlab/service/MqttBroker.java

```diff
@@ -0,0 +1,628 @@
+          }
+          send(msg);
+        }
+
+      }
+    } catch (Exception e) {
+      error(e);
+    }
+  }
+
+  @Override
+  public void onSubscribe(InterceptSubscribeMessage m) {
+    invoke("publishSubscribe", m);
+    try {
+      String topic = m.getTopicFilter();
+      if (topic != null && topic.length() > 0 && topic.charAt(0) != '$' && topic.charAt(0) != '/') {
+        topic = '/' + topic;
+      }
+
+      String[] parts = topic.split("/");
+
+      if (processApiMessages && parts.length == 5 && parts[1].equals(mrlTopicApiPrefix) && parts[2].equals("service")) {
+
+        subscribe(parts[3], parts[4], m.getClientID(), CodecUtils.getCallbackTopicName(parts[4]));
+        // subscribe(parts[2], parts[3], getName(), parts[3]);
+        // subscribe(parts[2], parts[3], m.getClientID(), parts[3]);
+      }
+    } catch (Exception e) {
+      error(e);
+    }
+
+    broadcastState();
+  }
+
+  @Override
+  public void onUnsubscribe(InterceptUnsubscribeMessage msg) {
+    invoke("publishUnsubscribe", msg);
+  }
+
+  @Override
+  public boolean preProcessHook(Message msg) {
+    // let the messages for this service
+    // get processed normally
+    if (methodSet.contains(msg.method)) {
+      return true;
+    }
+    // otherwise its target is for the
+    // scripting environment
+    // set the data - and call the call-back function
+    // handling call-back input needs to be
+    // done by another thread - in case its doing blocking
+    // or is executing long tasks - the inbox thread needs to
+    // be freed of such tasks - it has to do all the inbound routing
+
+    // FIXME - NOT NEEDED OR WANTED ?
+    MqttPublishMessage message = MqttMessageBuilders.publish().topicName(mrlTopicApiPrefix + "/" + msg.getFullName() + "/" + msg.getMethod()).retained(true)
+        .qos(MqttQoS.AT_LEAST_ONCE).payload(Unpooled.copiedBuffer(CodecUtils.toJson(msg).getBytes(UTF_8))).build();
+
+    mqttBroker.internalPublish(message, null); // I need the client id ? why?
+    return false;
+  }
+
+  public MqttMsg publishConnect(InterceptConnectMessage msg) {
+    return toMqttMsg(msg);
+  }
+
+  public MqttMsg publishConnectionLost(InterceptConnectionLostMessage msg) {
+    return toMqttMsg(msg);
+  }
+
+  public MqttMsg publishDisconnect(InterceptDisconnectMessage msg) {
+    return toMqttMsg(msg);
+  }
+
+  public MqttMsg publishMessageAcknowledged(InterceptAcknowledgedMessage msg) {
+    return toMqttMsg(msg);
+  }
+
+  public MqttMsg publishMqttMsg(MqttMsg msg) {
+    return msg;
+  }
+
+  public MqttMsg publishSubscribe(InterceptSubscribeMessage msg) {
+    return toMqttMsg(msg);
+  }
+
+  public MqttMsg publishUnsubscribe(InterceptUnsubscribeMessage msg) {
+    return toMqttMsg(msg);
+  }
+
+  private void saveIdentities() {
+    try {
+      if (username != null && username.length() > 0) {
+        FileOutputStream fos = new FileOutputStream(passwordFilePath);
+        MessageDigest digest = MessageDigest.getInstance("SHA-256");

```

Comment:
we should pull the algorithm into a constant at the top of the class.. or make it configurable with a default as provided here.

Reply:
that's icing .. but sure ... - the challenge was to get to work with a cloud provider

will fix

Reply:
fixed as constant - not configurable